### ISSUE TITLE
Campaigns 3.19 Tracking issue

### ISSUE THREAD

**Issue by mrnugget** [sourcegraph/sourcegraph#12328]
### Goal

**Release the [new spec-based campaigns workflow](https://docs.sourcegraph.com/@campaigns-new-flow/user/campaigns/hello_world_campaign) behind a feature flag for site-admins and non-site-admins.**

After Quinn designed the new campaigns workflow in the past two iterations in [#10921](https://github.com/sourcegraph/sourcegraph/pull/10921) we finished the API design and conceptual work in the last iteration and are now ready to implement the complete workflow. Since @eseliger and @mrnugget already started the implementation in 3.18, our goal for this iteration is to finish the implementation and release the new campaigns workflow to customers.

We are going to work on this single goal together as a team. @eseliger will be working on the frontend. After finishing GitLab support by implementing webhook support, @LawnGnome will first focus on the `src-cli` side and then join @mrnugget in implementing the missing pieces in the backend. We anticipate that the first half of the iteration will be spent working on well-defined tasks and the second half on polishing and bugfixing.

We also want to combine the release of the new workflow with the loosening of the campaigns feature-flag: instead of campaigns only being fully available to site-admins and in a read-only mode to users, we want to make keep the feature flag but, when enabled, make campaigns available to all users of a Sourcegraph instance.

Campaigns will still be marked as "beta", since we don't feel confident to go out of beta with the first release of the new workflow (that included major changes to large parts of our codebase). But the aim is that 3.19 will be the last release with campaigns as a beta feature and should be considered more of a "release candidate" than a beta. 

### Availability

Period is from July 20th to August 19th (23 working days). Please write the days you won't be working and the number of working days for the period.

**NOTE**: This iteration will also include Q3 planning and a new 360 review cycle.

- @mrnugget: 19d (on vacation: 12.-17. August)
- @eseliger: 23d
- @LawnGnome: 22d (BC Day 3rd August)

### Tracked issues

@LawnGnome: __7.00d__

- [ ] ~campaigns: GitLab integration tests~ ([#12219](https://github.com/sourcegraph/sourcegraph/issues/12219)) __2d__ 🧶

Completed: __7.00d__
- [x] (🏁 63 days ago) Add GitLab to the list of supported hosts ([#249](https://github.com/sourcegraph/src-cli/pull/249)) :shipit:
- [x] (🏁 57 days ago) schema: allow personal repos on Bitbucket Server ([#12504](https://github.com/sourcegraph/sourcegraph/pull/12504)) :shipit:🐛
- [x] (🏁 55 days ago) doc: make webhook configuration consistent ([#12529](https://github.com/sourcegraph/sourcegraph/pull/12529)) :shipit:
- [x] (🏁 50 days ago) campaigns: GitLab webhook support ([~#12171~](https://github.com/sourcegraph/sourcegraph/issues/12171)) __2d__ 
- [x] (🏁 40 days ago) Write a migration that deletes all existing campaigns/changesets when upgrading to 3.19 ([~#12641~](https://github.com/sourcegraph/sourcegraph/issues/12641)) 
- [x] (🏁 40 days ago) Allow customers to export their campaigns before upgrading to 3.19 ([~#12640~](https://github.com/sourcegraph/sourcegraph/issues/12640)) 
- [x] (🏁 40 days ago) campaigns: migrate to allow changeset spec user IDs to be NULL ([#13020](https://github.com/sourcegraph/sourcegraph/pull/13020)) :shipit:🐛
- [x] (🏁 36 days ago) campaigns: patch changeset IDs when migrating ([#13111](https://github.com/sourcegraph/sourcegraph/pull/13111)) :shipit:
- [x] (🏁 35 days ago) Add CLI option to update a campaign ([~#11350~](https://github.com/sourcegraph/sourcegraph/issues/11350)) 
- [x] (🏁 35 days ago) src-cli: Implement new campaigns workflow ([~#12333~](https://github.com/sourcegraph/sourcegraph/issues/12333); PRs: ~[#258](https://github.com/sourcegraph/src-cli/pull/258)~) __5d__ 

@Unassigned: __2.50d__

Completed: __2.50d__
- [x] (🏁 55 days ago) Implement API to return all involved unsers in a campaign ([~#12566~](https://github.com/sourcegraph/sourcegraph/issues/12566)) __1d__ 
- [x] (🏁 55 days ago) Implement campaignForSpec query resolver ([~#12567~](https://github.com/sourcegraph/sourcegraph/issues/12567)) __1d__ 
- [x] (🏁 48 days ago) Implement GraphQL API that allows to fetch campaigns by namespace ([~#12563~](https://github.com/sourcegraph/sourcegraph/issues/12563)) __0.5d__ 
- [x] (🏁 48 days ago) Migrate existing campaigns' names to the new slug-like format ([~#12649~](https://github.com/sourcegraph/sourcegraph/issues/12649)) 

@chrispine

Completed
- [x] (🏁 44 days ago) Remove Campaign.Branch field ([~#12735~](https://github.com/sourcegraph/sourcegraph/issues/12735)) 

@eseliger: __8.50d__

- [ ] ~Design a GraphQL API that allows us to filter changesets via the textbox~ ([#12471](https://github.com/sourcegraph/sourcegraph/issues/12471)) 

Completed: __8.50d__
- [x] (🏁 56 days ago) Decompose and/or add more tests for CampaignDetails ([~#10896~](https://github.com/sourcegraph/sourcegraph/issues/10896)) 🧶
- [x] (🏁 55 days ago) Design a GraphQL API that allows us to make use of the cached diff stats on ChangesetSpecs and Changesets ([~#12464~](https://github.com/sourcegraph/sourcegraph/issues/12464)) 
- [x] (🏁 55 days ago) Design a GraphQL API that allows us to list campaigns on user and organization pages ([~#12470~](https://github.com/sourcegraph/sourcegraph/issues/12470)) 
- [x] (🏁 55 days ago) Design the GraphQL API needed to show all users involved in a campaign ([~#12465~](https://github.com/sourcegraph/sourcegraph/issues/12465)) 
- [x] (🏁 51 days ago) Persist & accept changeset filter params in the URL ([~#12472~](https://github.com/sourcegraph/sourcegraph/issues/12472)) 
- [x] (🏁 49 days ago) ChangesetConnectionResolver.PageInfo needs to respect repository permissions ([~#12764~](https://github.com/sourcegraph/sourcegraph/issues/12764)) __0.5d__ 
- [x] (🏁 48 days ago) Design a GraphQL API that would allow us to show either the "Update campaign" or "Create campaign" button on the apply page ([~#12468~](https://github.com/sourcegraph/sourcegraph/issues/12468)) 
- [x] (🏁 38 days ago) Frontend: New campaigns flow ([~#10986~](https://github.com/sourcegraph/sourcegraph/issues/10986)) __8d__ 
- [x] (🏁 31 days ago) ~Replace "automation" feature flag with "campaigns" toggle and remove configuration flags~ ([~#10713~](https://github.com/sourcegraph/sourcegraph/issues/10713); PRs: ~[#11621](https://github.com/sourcegraph/sourcegraph/pull/11621)~, ~[#11621](https://github.com/sourcegraph/sourcegraph/pull/11621)~) __1d__ 🛠️
- [x] (🏁 today) ~campaigns: cancelled GitLab pipelines don't take check state out of pending~ ([~#12667~](https://github.com/sourcegraph/sourcegraph/issues/12667)) __1d__ 🐛

@mrnugget: __13.00d__

Completed: __13.00d__
- [x] (🏁 66 days ago) Another set of small follow-ups after introducing ExternalChangesetState ([#12330](https://github.com/sourcegraph/sourcegraph/pull/12330)) :shipit:
- [x] (🏁 65 days ago) Check namespace perms in CreateCampaignSpec/MoveCampaign ([#12362](https://github.com/sourcegraph/sourcegraph/pull/12362)) :shipit:
- [x] (🏁 64 days ago) Drop old campaigns tables and add missing columns ([#12388](https://github.com/sourcegraph/sourcegraph/pull/12388)) :shipit:
- [x] (🏁 51 days ago) Implement ApplyCampaign and changeset reconciler ([#12435](https://github.com/sourcegraph/sourcegraph/pull/12435)) :shipit:
- [x] (🏁 51 days ago) Backend: New campaign flow ([~#10985~](https://github.com/sourcegraph/sourcegraph/issues/10985); PRs: ~[#11675](https://github.com/sourcegraph/sourcegraph/pull/11675)~) __12d__ 
- [x] (🏁 51 days ago) Change testChangesetResponse in permissions_test.go to include publication status, error and reconciler state ([~#12647~](https://github.com/sourcegraph/sourcegraph/issues/12647)) 
- [x] (🏁 50 days ago) Creating larger campaigns using src-cli fails with 413 ([~#9202~](https://github.com/sourcegraph/sourcegraph/issues/9202)) __0.5d__ 🐛
- [x] (🏁 49 days ago) Implement the changesetsConnectionStatsResolver ([~#12642~](https://github.com/sourcegraph/sourcegraph/issues/12642); PRs: ~[#12753](https://github.com/sourcegraph/sourcegraph/pull/12753)~) 
- [x] (🏁 49 days ago) Implement missing methods in changesetResolver ([~#12646~](https://github.com/sourcegraph/sourcegraph/issues/12646); PRs: ~[#12753](https://github.com/sourcegraph/sourcegraph/pull/12753)~) 
- [x] (🏁 49 days ago) Replace commented-out portion of TestCampaigns in resolvers/resolver_test.go with more extensive test that tests ApplyCampaign ([~#12648~](https://github.com/sourcegraph/sourcegraph/issues/12648); PRs: ~[#12753](https://github.com/sourcegraph/sourcegraph/pull/12753)~) 
- [x] (🏁 49 days ago) Do not expire the PreviousChangesetSpec of a changeset ([~#12763~](https://github.com/sourcegraph/sourcegraph/issues/12763)) 
- [x] (🏁 48 days ago) Implement the "processing?" checks in CloseCampaign/DeleteCampaign ([~#12643~](https://github.com/sourcegraph/sourcegraph/issues/12643)) 
- [x] (🏁 45 days ago) Implement combined diff stat on changesetspec resolver ([~#12564~](https://github.com/sourcegraph/sourcegraph/issues/12564)) __0.5d__ 
- [x] (🏁 45 days ago) Change campaign/campaignSpec URLs to include namespace and fit new schema ([~#12473~](https://github.com/sourcegraph/sourcegraph/issues/12473)) 
- [x] (🏁 45 days ago) Implement the createCampaign mutation and necessary API fields ([~#12469~](https://github.com/sourcegraph/sourcegraph/issues/12469)) 
- [x] (🏁 44 days ago) Record enough information on a campaign to return all involved users in a campaign ([~#12467~](https://github.com/sourcegraph/sourcegraph/issues/12467)) 
- [x] (🏁 31 days ago) ~Replace "automation" feature flag with "campaigns" toggle and remove configuration flags~ ([~#10713~](https://github.com/sourcegraph/sourcegraph/issues/10713); PRs: ~[#11621](https://github.com/sourcegraph/sourcegraph/pull/11621)~, ~[#11621](https://github.com/sourcegraph/sourcegraph/pull/11621)~) __1d__ 🛠️
- [x] (🏁 22 days ago) ~Allow retrying of errored changesets~ ([~#12700~](https://github.com/sourcegraph/sourcegraph/issues/12700)) __4d__ 

#### Legend

- 👩 Customer issue
- 🐛 Bug
- 🧶 Technical debt
- 🛠️ [Roadmap](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit#heading=h.5nwl5fv52ess)
- 🕵️ [Spike](https://en.wikipedia.org/wiki/Spike_(software_development))
- 🔒 Security issue
- :shipit: Pull Request

**Comment by mrnugget** [comment]
@nicksnyder This tracking issue is now final. Including the goal. Please take a look :)

**Comment by nicksnyder** [comment]
LGTM, thanks!

**Comment by mrnugget** [comment]
**Weekly update: 2020-07-20 to 2020-07-24**

Last week we planned our 3.19 iteration (the goal and who's roughly doing what is described above in this issue).

We also had a retrospective and identified one thing that we need to improve: we need clearer ownership and clearer expectations. To address that, we set ourselves the goal of, when even slightly in doubt about whether the other person knows what to do, to create a ticket, with clear expectations and, if it exists, a timeline.

To help development, we also merged #11675, our WIP PR for the new workflow, into master, even though the workflow is not finished yet. But keeping the PR up to date and rebasing other PRs onto it turned into a burden. So on `master`, campaigns are now in a read-only mode that gets removed with more features and functionality being merged into `master`.

We also agreed on having two additional sync meetings per week to help us work together on this new workflow, by fixed time slots in which to sync, ask questions, identify blockers, etc.

I personally will focus solely on getting the reconciler and database layer for the new workflow working this week.

**Comment by mrnugget** [comment]
**Weekly update: 2020-07-27 to 2020-07-31**

No major updates on a team-level last week, except that our additional syncs are paying off — a lot. We had some really fruitful discussions around the design and architecture of the new reconciler and I think we should keep those additional sync meetings until the end of this iteration, at which I expect the new workflow to be completed on `master`.

I personally started the week trying to see whether I can get the changeset reconciler to work and ended it confident that the design works. There are still things left to do, such as adding integration tests, but the reconciler works, and it works well and is more extensible than the design we had before. I'm quite happy about that. [The PR that adds the reconciler and fleshes out the `ApplyCampaign` mutation](https://github.com/sourcegraph/sourcegraph/pull/12435) is open and has already been reviewed & approved by Adam and Erik. I'll add more tests and do some refactoring but plan to merge it in 1-2 days. After that I'm going to focus on finishing as many backend tickets as possible so that we can get the new workflow up and running as soon as we can.

On a team level this week is also quite exciting: we welcome @chrispine as the new engineering manager for campaigns! So we should all set aside some time to help Chris onboard and get to know Campaigns :)

**Comment by LawnGnome** [comment]
#### Weekly update: 2020-07-27 to 2020-07-31

Last week, I was optimistic that I would be done [GitLab webhooks](https://github.com/sourcegraph/sourcegraph/pull/12139) and could move on with my other big 3.19 tasks: most notably, [CLI support](https://github.com/sourcegraph/sourcegraph/issues/12333). [I was, sadly, wrong.](https://supercoolcreative.com/wp-content/uploads/2014/05/5-Things-Startups-Think-About-Themselves-That-Are-Wrong-supercool-creative.png)

However, we made good progress, if slower than I would have liked. Erik did a terrific job testing the GitLab webhook PR, and found a couple of bugs I'm _very_ glad we didn't ship. I also continued to bang my head against the inconsistencies between GitLab's webhooks and their other APIs. It's very [quirky](https://thesaurus.plus/img/synonyms/549/quirky.png).

Beyond that, I'd been having trouble with BitBucket Server and personal repos, so [I fixed that](https://github.com/sourcegraph/sourcegraph/pull/12504). I also put a fair amount of time into reviewing ongoing work. It feels like we're at a crucial point right now in shipping campaigns, between [Thorsten's work to implement the new reconciler state machine](https://github.com/sourcegraph/sourcegraph/pull/12435) and Erik's ongoing work on our frontend and GraphQL API, and I want to make sure that we're giving this important work the focus it deserves.

This week, after a public holiday Monday, it's full steam ahead on CLI design and implementation. Well, you know, just as soon as I merge that pesky GitLab webhook PR... :man_facepalming:

**Comment by eseliger** [comment]
#### Weekly update: 2020-07-27 to 2020-07-31

Last week, I did some manual test passes on GitLab webhook support and hence ruined Adams day. 
Other than that, I [brought the campaigns components back into shape (and the new types for GQL)](https://github.com/sourcegraph/sourcegraph/pull/12562) and fixed a ton of bugs that sneaked in when we ripped out all the previous workflow related components. Most of my remaining time was spent on discussions, reviews, GraphQL API designs and preparing for incorporating design into the feedback loop for the new UI this week.

**Comment by mrnugget** [comment]
**Weekly update: 2020-08-03 to 2020-08-07**

I made a lot of progress on finishing up the work on the new reconciler last week: added a lot of tests, finished the changeset resolvers, added defensive checks, etc. Basically I removed all the TODOs I put in in the two weeks before. Then I took on some more tickets to unblock Erik on the frontend and implemented missing resolvers and attributes he needs in the frontend. Other things that we didn't consider release blockers we moved to the backlog to get an overview of the things that definitely need to be done.

This week **I'll be on vacation from Wednesday to Monday next week**. But tomorrow I want to take a closer look at the reconciler and document its retry behaviour and make sure that it works.

**Comment by eseliger** [comment]
**Weekly update: 2020-08-03 to 2020-08-07**

I made progress on implementing more of the new campaigns components, and as a side-task to get to see something different than just storybooks was able to clean up our backend code a bit. I also had a design session together with Rob last week to get major blockers out of the way so I can focus this week on writing code all day so we have everything that's needed in place by branch cut, which will be my major task this week. I'll also probably spend some time talking through the src CLI interface with Adam and exchange further with Rob on the designs, especially the ones he came up with already.

**Comment by LawnGnome** [comment]
#### Weekly update: 2020-08-03 to 2020-08-07

Last week, I (finally) merged the GitLab webhook work, and moved on to the CLI work required by #12333. I'm tracking this [in a draft PR](https://github.com/sourcegraph/src-cli/pull/260), and feel fairly good about how that's going: I've removed the old commands, and am now building new commands using a mixture of the existing `internal/campaigns` code and the new functionality required to handle validating and processing the new campaign spec format.

Along the way, [I refactored the GraphQL client code in `src-cli` into a separate package](https://github.com/sourcegraph/src-cli/pull/258) to give us more options around organising code in `src-cli`: I don't really want to lump business logic into the `cmd/src` package willy-nilly, and this is the first step towards a more service-oriented, loosely coupled `src-cli`. [Definitely no master plan here, no sir...](https://giphy.com/gifs/vh1-love-and-hip-hop-lhhh-FfjiRWSBCpPiTJ2vaq)

This week, I'm implementing the changeset spec generation and landing the `src-cli` changes, then switching gears to work on the migration story for campaign users upgrading from 3.18 to 3.19, since we're making a bunch of breaking changes.

**Comment by chrispine** [comment]
**Weekly update: 2020-08-03 to 2020-08-07** 

Bottom line: *we are on track for cutting the 3.19 release branch this Friday*.

Biggest disruption to the team last week was getting a new manager! (Hopefully they won't be too much of a jerk.) In all seriousness, I did need some time from the team for onboarding, and which has continued this week. The negative impact to the team was only a couple of hours so far.

We took a hard look at our tracking issue items above, and bumped a few of the less urgent issues to 3.20, in order to ensure that we would be finished up by Friday. This was especially important given that Thorsten is taking some well-deserved time off (leaving Wed, back next Tue).

One pain point which surfaced was around the lack of an owner for the `src-cli` repo. I intend to take the matter up with the other EMs to see if other teams feel similarly, and to identify paths forward.

Personally, I finished one small ticket to remove the `Branch` field from Campaigns, now that they get the branch name from the spec. Most of the work was ironing out wrinkles in my setup. (Special thanks to Erik for figuring out that my tests were running *super*-slow because of my ISPs "helpful" DNS behavior.)

**Comment by marekweb** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.19 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by LawnGnome** [comment]
#### Weekly update: 2020-08-10 to 2020-08-14

Last week, I finished the initial work on [`src campaigns apply`](https://github.com/sourcegraph/src-cli/pull/260), which got us to the point where we could apply campaigns that have both created and imported changesets. Once that was done, I switched gears and worked on [the 3.18→3.19 migration story](https://github.com/sourcegraph/sourcegraph/pull/13001): campaigns that predate our new spec-based workflow will be migrated into campaigns with only imported changesets, which means that users will be able to track their existing campaigns, but will be unable to update them further. (Although I _do_ have an idea on that, if anyone really needs it.)

This week, I'll be switching back to the `src-cli` work: `src campaigns apply` is still missing an `-apply` flag to immediately apply campaigns, and its output and logging is fairly subpar right now:
![image](https://user-images.githubusercontent.com/229984/90421559-701a0800-e06e-11ea-9544-fd40a14e7f0a.png)

Once that's done, I'll be writing (or helping to write) our new campaign documentation for 3.19, and then starting to think about 3.20.

**Comment by eseliger** [comment]
#### Weekly update: 2020-08-10 to 2020-08-14

Last week, I've spent a lot of time with Design to work on the new campaigns UI and implemented as much as possible for that and we were finally able to try out campaigns as a whole from src-cli to creating actual pull requests on GitHub in on workflow after weeks of planning and implementing it, and I must admit it felt smooth and went without errors, so that gives me some good confidence. Besides, I've also fixes some errors that came up while testing, some of them were critical for 3.19 so we had to prioritize them over the campaign close UI, which I'm going to add this week and hopefully cherry-pick into 3.19.

So for this week, besides iteration planning and revisiting things that may have went a little too fast last week, I'll be on the close UI for the first half of the week and then spend the remaining time on refining the UI work or focus on bugs, when we find more.

**Comment by chrispine** [comment]
Last week took us in an exciting (if a bit unexpected) direction, with the new UI work that Erik mentions above. The work came together beautifully, but did make for a slightly more hectic end of week as the team hurried to get everything in before the cutting of the release branch. Great work all around!

Personally, I spent much of last week poking around in the code, and had my first learning session with Adam and Erik where I learned about the reconciler (and more about the implementation of campaigns in general). I have a personal goal of running an actual campaign on our actual codebase.

I also spent some time with Quinn talking about our team’s roadmap, potential milestones, and staying focused on the goal of getting everyone using campaigns (as opposed to getting hung up on beta vs GA). 

This week, I’m focusing on 3.20 planning prep and future milestone definitions. To that end, I will be reviewing our backlog, and meeting with the team to make sure we have alignment on our milestones. Then I will be adding that to our team goals.
